src/Database/FieldTypeConverter.php

```diff
@@ -49,20 +60,49 @@ public function __construct(TypeMap $typeMap, Driver $driver)
         $this->_driver = $driver;
         $map = $typeMap->toArray();
         $types = Type::buildAll();
-        $result = [];
+
+        $simpleMap = $batchingMap = [];
+        $simpleResult = $batchingResult = [];
 
         foreach ($types as $k => $type) {
             if ($type instanceof OptionalConvertInterface && !$type->requiresToPhpCast()) {
-                unset($types[$k]);
+                continue;
             }
+
+            if ($type instanceof BatchCastingInterface) {
```

Comment:
I think we have to add more methods which enables/disables the batch casting feature to the Type class:
```php
class Type implements TypeInterface
    static $batchCastingEnabled = [
        StringType::class => true,
        ...
    ];

    static function isBatchCastingEnabled($className) {
        return !empty(self::$batchCastingEnabled[$className]);
    }
    static function enableBatchCasting($className) { ... }
    static function disableBatchCasting($className) { ... }
}
```
```php
if ($type instanceof BatchCastingInterface && Type::isBatchCastingEnabled(get_class($type))) { 
    ...   
}
```

Otherwise, it will be hard to extend core Type classes. The following class no longer works in 3.next:
```php
class MyStringType extends StringType
{
    public function toPHP($value, Driver $driver) 
        // Do something
    }
}
```
Because toPHP() may not be called, in favor of manyToPHP(). People would have to copy&paste the manyToPHP() code of the parent class. Actually, we have to remove toPHP() from DateType and add $setToDateStart to the DateTimeType. 

If we allow people to enable/disable batch casting, the above MyStringType class will work in 3.next without any changes, as the class name is not in the $batchCastingEnabled array by default.

Reply:
I added $setToDateStart as an optimization to avoid calling an extra function, not to overcome any issues, but you may be right. In that this would make some methods not be called anymore.

Maybe document the change in the migration? I don’t think disabling the feature would help people not knowing that there was this change.

Reply:
Looks like your suggestion would work to prevent the issue you described. I’m just not too happy we have to do that.

Reply:
Is there an option where when batch casting is enabled, all casting operations are done through that method, even single columns? With that and an opt-in for batch casting will not cause problems in 3.x and in 4.x we could remove `toPHP` and always use batch casting.

Reply:
I agree with there is no need to disable the feature. I am OK if the above MyStringType works fine in 3.next without any changes.

Reply:
The problem is that using the batch when there is not batching to be done is a bit slower

Reply:
I'll measure again to see what's the downside of doing this

Reply:
I tried making `toPHP` always call `manyToPHP` internally for the `IntegerType` class. It resulted in a significant performance loss. So this is not the shortcut I'd like to try.

Another shortcut is to hardcode the batching only for core classes and classes not extending from Type. I think that would avoid having that enable/disable methods and would still preserve all of the benefits.

What do you think?

Reply:
We provide BatchCastingInterface only for users' classes, and core Types don't implement it? I think it works.